RavenDB-14510 Better controls over embedded server

ayende opened the pull request.

ayende committed:
RavenDB-14510 Better controls over embedded server

ppekrol commented on src/Raven.Embedded/EmbeddedServer.cs:
```diff
@@ -29,26 +31,48 @@ internal EmbeddedServer()
-            var startServer = new Lazy<Task<(Uri ServerUrl, Process ServerProcess)>>(() => RunServer(options));
-            if (Interlocked.CompareExchange(ref _serverTask, startServer, null) != null)
-                throw new InvalidOperationException("The server was already started");
+        public void Restart()
```
Maybe make this method async?

ppekrol replied:
It should be called `RestartServer` to match the `StartServer` one

ayende replied:
I made is sync and run in background because StartServer is also that.
But it make sense that you'll want to get any errors immediately.

ppekrol commented on src/Raven.Embedded/EmbeddedServer.cs:
```diff
@@ -193,13 +229,20 @@ private void ShutdownServerProcess(Process process)
         }
 
-        private async Task<(Uri ServerUrl, Process ServerProcess)> RunServer(ServerOptions options)
+        public event EventHandler? ServerProcessExited;
```
We should add `ServerProcessExitedEventArgs` here. They can be empty for now on, but it is easier to expand later on.

ayende committed:
RavenDB-14510  PR Comments

ppekrol committed:
RavenDB-14510
- we need to enable raising events
- async methods in public API needs to end with 'Async'
- added test

ppekrol committed:
RavenDB-14510 fixed NRE
